### PR TITLE
fix: Routing Forms, Number operators were using string operands

### LINKED ISSUE #9170: [CAL-1806] Routing Form: number comparison operators seem to operate on strings instead of numbers
Greater than operator b/w 500 and 1000 considers 500 as bigger.

[https://calcominc.slack.com/archives/C050R8Z2RP0/p1685143897762489](https://calcominc.slack.com/archives/C050R8Z2RP0/p1685143897762489)

<sub>From [SyncLinear.com](https://synclinear.com) | [CAL-1806](https://linear.app/calcom/issue/CAL-1806/routing-form-number-comparison-operators-seem-to-operate-on-strings)</sub>

### PR DESCRIPTION
**Pull request by hariombalhara**
## What does this PR do?
- Makes sure that number operators receives input as number instead of string.
- Implements missing >, >=, <, <= operators for Reporting.

Fixes #9170 Fixes CAL-1806

## Demo
[After the fix demo](https://www.loom.com/share/4f981c023e1b461a8aa862d995ff02f4)

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## How should this be tested?

- [x] Add a number type field say `age`.
	- Choose age greater than 50, and select router action as a custom page(You can set the custom page message as 'the age is greater than 50')
	- Go to Routing Form and input 100. Submit the form and you would see the Fallback Message instead when infact 100 > 50. This issue is fixed with the PR.
- [x] Similarly, for reporting, add a filter of age > 50 and check if you see the just added response with age 100. 
	- Now, add a filter of age < 50, you would not see the recently added response of age 50. But in the main you would see it.
- [x] Do a 'Test Preview' and make sure that adding a 100 their shows the custom page, 'the age is greater than 50'
## Checklist

- I haven't added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: packages/app-store/routing-forms/jsonLogicToPrisma.ts [hunk 1]
```diff
@@ -43,6 +43,22 @@ const OPERATOR_MAP = {
...
+    secondaryOperand: null,
+  },
+  "<": {
+    operator: "lt",
+    secondaryOperand: null,
+  },
+  "<=": {
+    operator: "lte",
+    secondaryOperand: null,
+  },
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/jsonLogicToPrisma.ts:61]: These operators were missing implementation

File: packages/app-store/routing-forms/pages/router/[...appPages].tsx [hunk 2]
```diff
@@ -60,12 +61,11 @@ export const getServerSideProps = async function getServerSideProps(
 
   const response: Record<string, Pick<Response[string], "value">> = {};
   serializableForm.fields?.forEach((field) => {
-    const rawFieldResponse = fieldsResponses[getFieldIdentifier(field)] || "";
-    console.log(field);
-    const fieldResponse =
-      field.type === "multiselect" ? rawFieldResponse.split(",").map((r) => r.trim()) : rawFieldResponse;
+    const fieldResponse = fieldsResponses[getFieldIdentifier(field)] || "";
+    const value =
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/pages/router/[...appPages].tsx:65]: Renamed fieldResponse as value(that better describes what it holds). Use, fieldResponse in place of rawFieldResponse.

File: packages/app-store/routing-forms/api/responses/[formId].ts [hunk 3]
```diff
@@ -39,7 +39,8 @@ async function* getResponses(
         if (value instanceof Array) {
           serializedValue = value.map((val) => escapeCsvText(val)).join(" | ");
         } else {
-          serializedValue = escapeCsvText(value);
+          // value can be a number as well for type Number field
+          serializedValue = escapeCsvText(String(value));
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/api/responses/[formId].ts:43]: value has type string|number here. Make it string for CSV.

File: packages/app-store/routing-forms/jsonLogicToPrisma.ts [hunk 4]
```diff
@@ -71,13 +88,55 @@ const convertSingleQueryToPrismaWhereClause = (
...
   }
-  const prismaWhere = {
-    response: { path: [mainOperand, "value"], [`${prismaOperator}`]: secondaryOperand },
-  };
+
+  const isNumberOperator = NumberOperators.includes(operatorName);
+  const secondaryOperandAsNumber = typeof secondaryOperand === "string" ? Number(secondaryOperand) : null;
+
+  let prismaWhere;
+  if (secondaryOperandAsNumber) {
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/jsonLogicToPrisma.ts:101]: I rely on `jsonLogicToPrisma.test.ts` here.
Have added 4 new tests for >,>=, <, <=

File: packages/app-store/routing-forms/pages/routing-link/[...appPages].tsx [hunk 5]
```diff
@@ -187,8 +187,10 @@ function getUrlSearchParamsToForward(response: Response, fields: NonNullable<Pro
       // If for some reason, the field isn't there, let's just
       return;
     }
+    const valueAsStringOrStringArray =
+      typeof fieldResponse.value === "number" ? String(fieldResponse.value) : fieldResponse.value;
```
Comment (hariombalhara) [inline packages/app-store/routing-forms/pages/routing-link/[...appPages].tsx:191]: fieldResponse.value is number|string|string[] here. Make it string|string[]

File: packages/types/AppGetServerSideProps.d.ts [hunk 6]
```diff
@@ -1,7 +1,7 @@
-import { GetServerSidePropsContext, GetServerSidePropsResult } from "next";
-import { CalendsoSessionUser } from "next-auth";
+import type { GetServerSidePropsContext, GetServerSidePropsResult } from "next";
```
Comment (hariombalhara) [inline packages/types/AppGetServerSideProps.d.ts:1]: Prettier did it.

**Review by hariombalhara — COMMENTED** [review]
Self review done

**Review by zomars — APPROVED** [review]
Tested and it's working. Nice job @hariombalhara 

LGTM

**Pull request merged**